tests/exporter/collectors/test_cpu.py

```diff
@@ -36,18 +36,23 @@ def test_temp(self):
         else:
             self.assertIsNone(temp)
 
-    """
-    TODO: this test doesn't always pass.
+
+    # TODO: this test doesn't always pass.
     def test_str(self):
+        """
         @mccooeyc11
         Test whether the string representation of CPU metrics is correct
+        """
         cpu_collector = CPUCollector()
         utilization = cpu_collector.get_utilization()
         frequency = cpu_collector.get_frequency()
-        temp = cpu_collector.get_temperature().__str__()
-        expected_output = f"CPU utilization: {utilization}%\n CPU frequency: {frequency}MHz\n CPU temperature: {temp}°C\n"
+        temp = cpu_collector.get_temperature()
+        if temp:
+            temp = temp.__str__()
+            expected_output = f"CPU utilization: {utilization}%\n CPU frequency: {frequency}MHz\n CPU temperature: {temp}°C\n"
+        else:
+            expected_output = f"CPU utilization: {utilization}%\n CPU frequency: {frequency}MHz\n CPU temperature: not available\n"
         self.assertEqual(str(cpu_collector), expected_output)
```

Comment:
@ljdzed was talking about refactoring the test_str method to not call the functions from the CPUCollector().
There are two main issues about how the test is written right now:

**Problem 1**
The point of unit testing is to test a single "unit" of code. The way in which the test is written right now is actually testing multiple things because its calling all the different get functions and verifying the right values are returned.
What we should do instead is verify the formatting of the string.
You can discuss more about this in the backend meeting.

**Problem 2**
The reason why the tests were failing is because when we call `.get_frequency()` at line 48 and then call `str(cpu_collector)` at line 55, the value has a small chance to have changed between the two calls.

Reply:
One way to approach this might be to check if use regex or to check that terms like "CPU utilization: " and "CPU frequency: " appear in the response string along with "MHz" etc using some from of *string contains* or its equivalent in python.